modules/assim.batch/R/pda.emulator.ms.R

```diff
@@ -0,0 +1,518 @@
+      
+      ## re-fit GP on new param space
+      for(i in seq_along(SS.stack)){
+        GPmodel <- lapply(SS.stack[[i]], function(x) mlegp::mlegp(X = x[, -ncol(x), drop = FALSE], Z = x[, ncol(x), drop = FALSE], nugget = 0, nugget.known = 1, verbose = 0))
+        gp.stack[[i]] <- GPmodel
+      }
+      
+      ## re-define rng
+      rng <- norm_transform$rng
+      
+      ## get new init.list and jmp.list
+      init.list <- norm_transform$init
+      jmp.list  <- norm_transform$jmp
+
+      
+    }
+    
+
+    ########### hierarchical MCMC function with Gibbs ##############
+    
+    
+    hier.mcmc <- function(settings, gp.stack, nstack, nmcmc, rng, 
+                          mu0, jmp0, nparam, nsites){
+
+      pos.check <- sapply(settings$assim.batch$inputs, `[[`, "ss.positive")
+      
+      if(length(unlist(pos.check)) == 0){
+        # if not passed from settings assume none
+        pos.check <- rep(FALSE, length(settings$assim.batch$inputs))
+      }else if(length(unlist(pos.check)) != length(settings$assim.batch$inputs)){
+        # maybe one provided, but others are forgotten
+        # check which ones are provided in settings
+        from.settings <- sapply(seq_along(pos.check), function(x) !is.null(pos.check[[x]]))
+        tmp.check <- rep(FALSE, length(settings$assim.batch$inputs))
+        # replace those with the values provided in the settings
+        tmp.check[from.settings] <- as.logical(unlist(pos.check))
+        pos.check <- tmp.check
+      }else{
+        pos.check <- as.logical(pos.check)
+      }
+      
+      ################################################################
+      #
+      #      mu_site    : site level parameters (nsite x nparam)
+      #      tau_site   : site level precision (nsite x nsite)
+      #      mu_global  : global parameters (nparam)
+      #      tau_global : global precision matrix (nparam x nparam)
+      #
+      ################################################################
+      
+      
+      
+      ###### (hierarchical) global mu priors
+      #
+      #      mu_f    : prior mean vector
+      #      P_f     : prior covariance matrix
+      #      P_f_inv : prior precision matrix
+      #
+      #      mu_global ~ MVN (mu_f, P_f)
+      #
+
+      mu_f    <- unlist(mu0)
+      P_f     <- diag(jmp0)
+      P_f_inv <- solve(P_f) 
+      
+      # # initialize mu_global (nparam)
+      repeat{
+        mu_global <- mvtnorm::rmvnorm(1, mu_f, P_f)
+        check.that <- (mu_global > apply(rng,1:2,max)[, 1] & mu_global < apply(rng,1:2,min)[, 2])
+        if(all(check.that)) break
+      }
+      
+      ######  (hierarchical) global tau priors
+      #
+      #      tau_df    : Wishart degrees of freedom
+      #      tau_V     : Wishart scale matrix
+      #
+      #      tau_global ~ W (tau_df, tau_scale)
+      #      sigma_global <- solve(tau_global)
+      #
+      
+      tau_df <- nsites + nparam + 1
+      tau_V  <- diag(1, nparam)
```

Comment:
yeah, tau_df is only used once to initialize tau_global which is probably unnecessary because tau_global is immediately updated in the loop. That's why I initialized it to `nsites + nparam + 1` 

some of these initializations were to wrap my head around dimensions, hyperparameters etc. I'll try to document it more